### ISSUE TITLE
Long-running progress notifications slow down neovim

### ISSUE THREAD

**Issue by vogelsgesang** [rcarriga/nvim-notify#71]
**Observed Issue**
If a LSP progress notification is open for too long, nvim-notify and neovim itself become sluggish.
I noticed this today while `clangd` was indexing a rather large code base (which meant the progress notification was open for multiple minutes)

Can be reproduced by running the following simplified script and keeping the vim open for 10 minutes. As you will see, the spinner will become laggy, and also using neovim in general will have a lot of lag.

```
local spinner_frames = { "⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷" }

local function update_spinner(notif_data)
   local new_spinner = (notif_data.spinner + 1) % #spinner_frames
   notif_data.spinner = new_spinner

   notif_data.notification = vim.notify(nil, nil, {
     hide_from_history = true,
     icon = spinner_frames[new_spinner],
     replace = notif_data.notification,
   })

   vim.defer_fn(function()
     update_spinner(notif_data)
   end, 100)
end

notif_data = {}
notif_data.notification = vim.notify("Test", "info", {
  title = "Leave this open for a while...",
  icon = spinner_frames[1],
  timeout = false,
  hide_from_history = false,
})
notif_data.spinner = 1

update_spinner(notif_data)
```

**My guess what the root cause could be:**
It seems we keep a history of all `hide_from_history` events, and just hide them when displaying the history.
This means, that if a notification with a spinner (which posts 10 hidden messages per second) stays open for too long, all the hidden notification events will grind nvim-notify and neovim itself to a halt.

**Comment by rcarriga** [comment]
Yeah something seemed off, I tested your script and saw issues with only a few hundred messages. Lua tables are very efficient and the data per notification is tiny. The only time the list of notifications is iterated over is for the history function. 

Tracked the actual cause for the slowdown to the extmarks for the buffer not being cleared. Now that they're cleared I'm easily hitting 11000 messages with no noticeable performance difference. Only stopped there because I got bored :sweat_smile:
